derive/src/decode.rs

```diff
@@ -300,3 +285,44 @@ fn create_instance(
 		},
 	}
 }
+
+pub fn quote_decode_with_mem_tracking(data: &Data, crate_path: &syn::Path) -> TokenStream {
+	let fields: Box<dyn Iterator<Item = &Field>> = match data {
+		Data::Struct(data) => Box::new(data.fields.iter()),
+		Data::Enum(ref data) => {
+			let variants = match utils::try_get_variants(data) {
+				Ok(variants) => variants,
+				Err(e) => return e.to_compile_error(),
+			};
+
+			let mut fields: Box<dyn Iterator<Item = &Field>> = Box::new(iter::empty());
+			for variant in variants {
+				fields = Box::new(fields.chain(variant.fields.iter()));
+			}
+			fields
+		},
+		Data::Union(_) => {
+			return Error::new(Span::call_site(), "Union types are not supported.")
+				.to_compile_error();
+		},
+	};
+
+	let processed_fields = fields.filter_map(|field| {
+		if utils::should_skip(&field.attrs) {
+			return None;
+		}
+
+		let field_type = if let Some(compact) = utils::get_compact_type(field, crate_path) {
+			compact
+		} else if let Some(encoded_as) = utils::get_encoded_as_type(field) {
+			encoded_as
+		} else {
+			field.ty.to_token_stream()
+		};
+		Some(quote! {<#field_type as #crate_path::DecodeWithMemTracking>::__is_implemented();})
```

Comment:
This `__is_implemented` function should be removed and the checks should be put into the unnamed `const` and using something like this:
https://docs.rs/static_assertions/1.1.0/src/static_assertions/assert_impl.rs.html#57-83

(We should copy this and not import the crate) Generally we could probably also have the following code:
```
fn require_decode_with_mem::<T: DecodeWithMemTracking>() {}

require_decode_with_mem::<field_ty>()
```

This should probably also work.

Reply:
Nice ! Done.